docs(rfc): create file naming guidelines for convergence (#16870)
microsoft/fluentui: rfcs/convergence/file-naming.md

<!-- Explain the proposed change -->

Apply consistent, easy to navigate style guide for naming files within converged components.

## Background

<!-- If there is relevant background include it here -->

This RFC builds on [RFC for convergence implementation patterns](https://github.com/microsoft/fluentui/pull/16806/files).

## Problem statement

<!--
Why are we making this change? What problem are we solving? What do we expect to gain from this?

This section is important as the motivation or problem statement is indepenent from the proposed change. Even if this RFC is not accepted this Motivation can be used for alternative solutions.

In the end, please make sure to present a neutral Problem statement, rather than one that motivates a particular solution
-->

Currently converged package contains following files:

> We'll use contrived name of package - `example` which lives in `/packages/example` folder

- `index.ts` - barrel / Public API surface
- `Example.tsx` - (complete component to be used as is)
- `useExample.ts` - (CONTROLLER/MODEL /main logic behind component - re-usable hook)
- `useExampleClasses.ts` - (STYLES/ main logic for styling component - re-usable hook)

> - `useExampleRandom.tsx` - (some encapsulated logic - not really a hook)

Reviewer: Hmm, could we then maybe put this inside a `useExample` dir to signify that it's additional encapsulated logic? Would that make sense?

Reviewer: Oh wait, this is what we currently have. Nvm.

- `renderExample.tsx` - (VIEW/ main logic that handles rendering - re-usable hook)
- `Example.types.ts` - (TYPES/ all TypeScript types used within package)

**Potential issues from developer point of view:**

- inconsistent file naming
  - some files contain suffix (`.types`)
  - some files have `.ts` some `.tsx` what's the difference ?
  - some files are PascalCased, some camelCased
- not explicitly communicating the intent
  - what does `use` mean ?

>     - seasoned react dev will probably guess that it's a hook. Thing is, that this file might export more that 1 hook.

Reviewer: That's a great catch!

## Detailed Design or Proposal

<!-- This is the bulk of the RFC. Explain the proposal or design in enough detail for the inteded audience to understand. -->

- use consistent names for all files.

> - follow a pattern that describes the symbol's feature or the file scope(component name) then its type.

Reviewer: ```suggestion
- follow a pattern that describes the symbol's feature or the file scope (component name) followed by its type.
```

  - The recommended pattern is `feature.type.ts`.
- use conventional **type** names including
  - `.hook`
  - `.view`
  - `.styles`
  - `.types`
  - `.utils`
  - `.stories`
  - invent additional type names if you must but take care not to create too many

**Why?**

- Type names provide a consistent way to quickly identify what is in the file.
- Type names make it easy to find a specific file type using an editor or IDE's fuzzy search techniques.
- Type names provide pattern matching for any automated tasks.

---

- use `.tsx` only for files that contain view/rendering related logic

**Why?**

- will enforce to use JSX only where necessary (renderer only)
- same as for Type Names

---

- (optional) use `.spec.` instead `.test.` suffix as we already follow behavioral driven approach (describes/it) rather than atomic `test()` approach
  - as a next step, enable lint rule that will enforce one way or another within tests

**Why?**

- consistency/proper intent

---

- (optional) use kebab case for naming

**Why?**

> - kebab case is more human readable especially in longer words

Reviewer: Either kebab or snake, but agreed that either is superior to camel or Pascal case for file names.

> - mitigates issues with git on various OS/case in/consistent files systems (eg. rename `helloWorld` to `helloworld`, commit, push, merge... now someone wants to change it back to `helloWorld`... )

Reviewer: Is JS/TS not agnostic to this issue? That is, are these not equivalent:

```ts
import { anApple } from './anApple';
```

and

```ts
import { anApple } from './anapple';
```

Author: it's all about file system case sensitivity AFAIK (if your os is case insensitive, both are valid)

Reviewer: FWIW, to the best of my knowledge, it is mainly macOS which is weird in this space and by default has case-insensitive filesystem. Both ext4 and ntfs by default are case-sensitive.

| ------------------- | ------------------- | ----------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| index.tsx           | index.ts            |                                                                                     | No changes from current                                                                                                                                                                                                                                                                                                   |

> | Example.tsx         | example.ts          | `export const Example = () => {..}`                                                 | To enforce rendering logic only within `.view.` type file/s. TS will enforce this by default. Not extremely picky on this one, we can use `.tsx` as well to match extension with test/story - priority is to be consistent                                                                                                |

Reviewer: Personally, I find the naming `Example.tsx` to be intuitive since it tells me we're exporting an object/component/etc. In fact, this is how the naming convention works in Zig as well for instance - if a module is in fact a struct, then name it as capital such as `Example`; otherwise, lowercase.

Author: I hear you. Just note that this is primarily about having consistency.

| Example.types.ts    | example.types.ts    | `export type Foo = {}; export interface Hello {}`                                   | aggregated set of reusable type definitions used within other parts of component/also might be re-exported for public API surface                                                                                                                                                                                         |

### Explanation

- use consistent names for all files.
- follow a pattern that describes the symbol's feature or the file scope(component name) then its type.
  - The recommended pattern is `feature.type.ts`.
- use conventional **type** names including
  - `.hook`
  - `.view`
  - `.styles`
  - `.types`
  - `.stories`
  - `.utils`
  - invent additional type names if you must but take care not to create too many

**Why?**

- Type names provide a consistent way to quickly identify what is in the file.
- Type names make it easy to find a specific file type using an editor or IDE's fuzzy search techniques.
- Type names provide pattern matching for any automated tasks.

---

> - use `.tsx` only for files that contain view/rendering related logic

Reviewer 2: ```suggestion
- use `.tsx` only for files that contain view/rendering related logic
  - enable lint rule that will enforce one way 
```

I suggest to add into proposal an ESLint rule for this, otherwise it's easy to fail this convention. There is [jsx-filename-extension](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-filename-extension.md), but I have not seen a similar for TS files:
- no JSX in files - use `.ts`
- JSX in files - use `.tsx` (this is forced by compiler anyway)

**Why?**

- will enforce to use JSX only where necessary (renderer only)
- same as for Type Names

---

- (optional) use `.spec.` instead `.test.` suffix as we already follow behavioral driven approach (describes/it) rather than atomic `test()` approach
  - as a next step, enable lint rule that will enforce one way or another within tests

**Why?**

- consistency/proper intent

---

- (optional) use kebab case for naming

**Why?**

- kebab case is more human readable especially in longer words

> - mitigates issues with git on various OS/case in/consistent files systems (eg. rename `helloWorld` to `helloworld`, commit, push, merge... now someone wants to change it back to `helloWorld`... )

Reviewer 2: This is handled by TypeScript compiler via `forceConsistentCasingInFileNames` (https://www.typescriptlang.org/docs/handbook/compiler-options.html) and is enabled across the repo. Is it a real currently?

-->

---

_List contributors to the proposal : @hotell_

## Summary

<!-- Explain the proposed change -->

Apply consistent, easy to navigate style guide for naming files within converged components.

## Background

<!-- If there is relevant background include it here -->

This RFC builds on [RFC for convergence implementation patterns](https://github.com/microsoft/fluentui/pull/16806/files).

## Problem statement

<!--
Why are we making this change? What problem are we solving? What do we expect to gain from this?

This section is important as the motivation or problem statement is indepenent from the proposed change. Even if this RFC is not accepted this Motivation can be used for alternative solutions.

In the end, please make sure to present a neutral Problem statement, rather than one that motivates a particular solution
-->

Currently converged package contains following files:

> We'll use contrived name of package - `example` which lives in `/packages/example` folder

- `index.ts` - barrel / Public API surface
- `Example.tsx` - (complete component to be used as is)
- `useExample.ts` - (CONTROLLER/MODEL /main logic behind component - re-usable hook)

> - `useExampleClasses.ts` - (STYLES/ main logic for styling component - re-usable hook)

Reviewer: FWIW I think the current convention is to name it `useExampleStyles.ts`.

| ------------------- | ------------------- | ----------------------------------------------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| index.tsx           | index.ts            |                                                                                     | No changes from current                                                                                                                                                                                                                                                                                                   |

> | Example.tsx         | example.ts          | `export const Example = () => {..}`                                                 | To enforce rendering logic only within `.view.` type file/s. TS will enforce this by default. Not extremely picky on this one, we can use `.tsx` as well to match extension with test/story - priority is to be consistent                                                                                                |

Reviewer 2: ```suggestion
| Example.tsx         | Example.ts          | `export const Example = () => {..}`  |
```

- it's a common pattern to use Pascal Case for components: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-pascal-case.md
- it's a common pattern to match filename to default export:
  - https://github.com/mui-org/material-ui/tree/next/packages/material-ui/src/Accordion
  - https://github.com/adobe/react-spectrum/tree/main/packages/%40react-spectrum/alert/src
  - https://github.com/reakit/reakit/tree/master/packages/reakit/src/Box
  - https://github.com/Semantic-Org/Semantic-UI-React/tree/master/src/elements/Button
  - https://github.com/segmentio/evergreen/tree/master/src/avatar/src
- React team uses Pascal Case in docs https://reactjs.org/docs/code-splitting.html#reactlazy

Author: I provided explicit reasoning in this document. This kind of feedback is not very helpful TBH.

- we are not react team
- default exports are an anti-pattern (nor is fluent using that ECMA feature)
- we can start passing links indefinitely, to showcase who uses what :)

Reviewer 3: I agree with Shift on this one, I'd prefer to keep the uppercase filename basically for the reasons he listed. 
- Clearly indicates that this file exports a component (even if it doesn't contain any rendering logic)
- Matches the name of the main/only export (not relevant for this purpose whether that export is default or named)
- Matches the convention used by both v8 and v0 today, and is extremely common throughout the community. Even though nobody has declared this as the "official" convention, at least from briefly looking through other React component repos, it's harder to find a repo that *doesn't* use uppercase names for component files than one that does.

> | Example.test.tsx    | example.spec.tsx    | `describe->it`                                                                      | Integration test on JSDOM level (jest) for whole component                                                                                                                                                                                                                                                                |

Reviewer 2: ```suggestion
| Example.test.tsx    | Example.spec.tsx    | `describe->it`                                                                      | Integration test on JSDOM level (jest) for whole component                                                                                                                                                                                                                                                                |
```

If we will go with Pascal Case, this should be also in Pascal Case: https://github.com/microsoft/fluentui/pull/16870/files#r574385390

| storybook (NONE)    | example.stories.tsx | `export const Example = () => {..}`                                                 | Specific file type for Storybook consumption. Main "starting" point for inner loop - daily development.                                                                                                                                                                                                                   |

> | mdx (NONE)          | example.stories.mdx | ---                                                                                 | Markdown(MDX) for consumer facing documentation(mainly for storybook at the moment, but can be transformed to any other tool that can parse MDX).                                                                                                                                                                         |

Author: you mean with `mdx` or the table in general?

> | mdx (NONE)          | example.stories.mdx | ---                                                                                 | Markdown(MDX) for consumer facing documentation(mainly for storybook at the moment, but can be transformed to any other tool that can parse MDX).                                                                                                                                                                         |

Reviewer 4: stories.mdx files 😄

> | mdx (NONE)          | example.stories.mdx | ---                                                                                 | Markdown(MDX) for consumer facing documentation(mainly for storybook at the moment, but can be transformed to any other tool that can parse MDX).                                                                                                                                                                         |

Author: haha, yeah so, again just a showcase to point out on other files "naming" (that we probably plan to use but are not used in current state). that's why I put there ` (NONE)`.

Is it confusing? WDYT ?

> | mdx (NONE)          | example.stories.mdx | ---                                                                                 | Markdown(MDX) for consumer facing documentation(mainly for storybook at the moment, but can be transformed to any other tool that can parse MDX).                                                                                                                                                                         |

Reviewer 4: While I see the value, it is slightly confusing if we've yet to use it as it felt more of a "this is the new standard".
What do you think of having a separate table just to create nomenclatures for possible future usages?

> | mdx (NONE)          | example.stories.mdx | ---                                                                                 | Markdown(MDX) for consumer facing documentation(mainly for storybook at the moment, but can be transformed to any other tool that can parse MDX).                                                                                                                                                                         |

Author: excellent idea, thx

> | renderExample.tsx   | example.view.tsx    | `export const render = () => {..}`                                                  | `.render` might work as well, to match symbol name.                                                                                                                                                                                                                                                                       |

Reviewer 2: A function should not be named as `render`, it should be `renderComponentName` as we should avoid collisions in consumer imports. For example:

```jsx
import { render } from 'react-dom'
import { render } from '@fluentui/react-button'
import { render } from '@fluentui/react-menu'
```

It also breaks DX on our side as if search will be used how many matches for `render` you will get?

Reviewer 2: For me, it's confusing to name files as `example.view.tsx` when a single export is `renderExample` function. It's not a common thing in React community:
- https://github.com/mui-org/material-ui/tree/next/packages/material-ui-utils/src
- https://github.com/adobe/react-spectrum/tree/main/packages/%40react-spectrum/utils/src

Author: > `renderComponentName`

That's a typo, thx.

> A function should not be named as `render`, it should be `renderComponentName` as we should avoid collisions in consumer imports. For example:

FWIW that's not true :) consumer can name imported symbols anyhow he wants. 

> It also breaks DX on our side as if search will be used how many matches for render you will get?

maybe out of topic but my understanding was that those renderers are meant to be used solely per package/component

Reviewer 5: "consumer can name imported symbols anyhow he wants." - from my experience with navigating relatively large code bases - the more specificity you have in names and the more consistency there is between file and export names, the better - because then you can easily combine fulltext search and file name search

In 99% of the cases probably yes, renderers will be used only in the package and component. But if the consumer wants to extend the component and add/replace hooks in it, he can still import and reuse the `render` function.

> | useExample.ts       | example.hook.ts     | `export const useExample = () => {}`                                                | main logic encapsulated under hook -> name should be the same as encapsulated component (in our case `example`)                                                                                                                                                                                                           |

Reviewer 2: This breaks common convention in React community:
- https://github.com/adobe/react-spectrum/tree/main/packages/%40react-spectrum/utils/src
- https://github.com/reakit/reakit/tree/master/packages/reakit-utils/src
- https://github.com/mui-org/material-ui/tree/next/packages/material-ui-utils/src
- https://github.com/streamich/react-use/tree/master/src

Author: I'd just add that there are no existing conventions ratified in any kind of official community style-guide, nor in fluent ui org :)  that's the primary aim of this RFC -> focus/ratification on consistency.

Reviewer 2: That's fair, I used these repos as a reference to show a common pattern. Fluent's repo follows it, too, currently.
Another fair note will be, that there is an explicit naming convention for hooks, they should start with `use*()` keyword.

https://github.com/facebook/react/blob/483358c38f8623358dd44a3e57476bc02e6357bc/packages/eslint-plugin-react-hooks/src/RulesOfHooks.js#L17-L19

| useExampleStyles.ts | example.styles.ts   | `export const useStyles = makeStyles(); export const useIconStyles = makeStyles();` | with `makeStyles` 1 to N hooks for particular component styling needs. Why we currently use a solution that provides hooks, having `.styles` could be a good enough convention to explicitly communicate the intent. (if we switch implementation in the future we wouldn't need to change the type name as it's generic) |

> | useExampleRandom.ts | example.utils.ts    | `export const foo = () => {}`                                                       | aggregated set of reusable utils used within other parts of component implementation. Note there are no real react hooks!                                                                                                                                                                                                 |

Author: we have something like this: https://github.com/microsoft/fluentui/tree/master/packages/react-avatar/src/components/utils

I added it rather for demonstration if such a case would occur. I can remove it from the RFC if you think it might be confusing ?

> | useExampleRandom.ts | example.utils.ts    | `export const foo = () => {}`                                                       | aggregated set of reusable utils used within other parts of component implementation. Note there are no real react hooks!                                                                                                                                                                                                 |

Reviewer 4: Don't think it's confusing at all, mostly just not a fan of utils 😄. I think it's best if we keep it for edge cases but mark it as optional, what do you think?

> | useExampleRandom.ts | example.utils.ts    | `export const foo = () => {}`                                                       | aggregated set of reusable utils used within other parts of component implementation. Note there are no real react hooks!                                                                                                                                                                                                 |

Author: > mark it as optional

what do you mean ?

> | useExampleRandom.ts | example.utils.ts    | `export const foo = () => {}`                                                       | aggregated set of reusable utils used within other parts of component implementation. Note there are no real react hooks!                                                                                                                                                                                                 |

Reviewer 4: My intention was to somehow make sure everyone is aware .utils files are optional, to prevent the idea that one would need all these files regardless of implementation.

| Example.types.ts    | example.types.ts    | `export type Foo = {}; export interface Hello {}`                                   | aggregated set of reusable type definitions used within other parts of component/also might be re-exported for public API surface                                                                                                                                                                                         |

### Explanation

- use consistent names for all files.
- follow a pattern that describes the symbol's feature or the file scope(component name) then its type.
  - The recommended pattern is `feature.type.ts`.
- use conventional **type** names including
  - `.hook`
  - `.view`
  - `.styles`
  - `.types`
  - `.utils`

>   - `.stories`

Author: same here. My thinking was following:
- `view`: we will export 1 renderer for 1 particular component (1:1)
- `hook`: every hook should live in separate file (main state/behaviour/prop hook `useExample` , a11y hook `useA11yExample()`, custom state hook (using manual `useState/Reducer()` etc )  

- rest is plural as usually (I'd assume that) there tends to be more than 1 symbol in those files

>   - `.stories`

Reviewer 4: Sounds good to me. In it's essence, if we export more than one thing, we pluralize it?

>   - `.stories`

Author: that sounds like good enough rule to me 👍
